crates/amaru-consensus/src/consensus/store.rs

```diff
@@ -37,8 +39,152 @@ impl Display for StoreError {
+    UnknownHeader { header: Hash<32> },
+
+    #[error("no parent header for: {header} (where one is clearly expected)")]
+    NoParentHeader { header: Hash<32> },
+
+    #[error("{0}")]
+    StoreError(#[from] StoreError),
+}
+
+impl<H: IsHeader> Praos<H> for dyn ChainStore<H> {
+    type Error = NoncesError;
+
+    fn get_nonce(&self, header: &Hash<32>) -> Option<Nonce> {
+        self.get_nonces(header)
+            .map(|(_epoch, nonces)| nonces.active)
+    }
+
+    fn evolve_nonce(&mut self, header: &H) -> Result<(), Self::Error> {
+        let (epoch, is_within_stability_window) = nonce::randomness_stability_window(header);
+
+        let parent_hash = header.parent().ok_or_else(|| NoncesError::NoParentHeader {
+            header: header.hash(),
+        })?;
+
+        let (parent_epoch, parent) =
+            self.get_nonces(&parent_hash)
+                .ok_or_else(|| NoncesError::UnknownParent {
+                    header: header.hash(),
+                    parent: parent_hash,
+                })?;
+
+        // Compute the new evolving nonce by combining it with the current one and the header's VRF
+        // output.
+        // Specifically, we combine it with `η` (a.k.a eta), which is a blake2b-256 hash of the
+        // tagged leader VRF output after a range extension. The range extension is, yet another
+        // blake2b-256 hash.
+        let evolving = nonce::evolve(header, &parent.evolving);
+
+        self.put_nonces(
+            &header.hash(),
+            epoch,
+            Nonces {
+                evolving,
+
+                // On epoch changes, compute the new active nonce by combining:
+                //   1. the (now stable) candidate; and
+                //   2. the previous epoch's last block's parent header hash.
+                //
+                // If the epoch hasn't changed, then our active nonce is unchanged.
+                active: if epoch > parent_epoch {
+                    let previous_epoch_last_header = self
+                        .load_header(&parent.previous_epoch_last_header)
+                        .ok_or(NoncesError::UnknownHeader {
+                            header: parent.previous_epoch_last_header,
+                        })?;
+                    nonce::from_candidate(&previous_epoch_last_header, &parent.candidate).ok_or(
+                        NoncesError::NoParentHeader {
+                            header: parent.previous_epoch_last_header,
+                        },
+                    )?
+                } else {
+                    parent.active
+                },
+
+                // Unless we are within the randomness stability window, we also update the candidate. This
+                // means that outside of the stability window, we always have:
+                //
+                //   evolving == candidate
+                //
+                // They only diverge for the last blocks of each epoch; The candidate remains stable while
```

Comment:
Rather than rather long comment, would it not make sense to have unit or property tests here?

Reply:
A pure tested function containing all the possible transitions for the nonce would seem to be more readable and maintainable, as there a somewhat significant of cases we want to ensure covered/